zkevm-circuits/src/taiko_pi_circuit.rs

```diff
@@ -417,15 +1284,1206 @@ impl<F: Field> TaikoPiCircuitConfig<F> {
+                                }),
+                            ),
+                        )
+                    })
+                    .collect_vec()
+                    .as_slice(),
+            );
+            block_data.extend_from_slice(
+                block_values
+                    .history_hashes
+                    .iter()
+                    .enumerate()
+                    .map(|(i, h)| {
+                        (
+                            "prev_hash lo",
+                            BlockContextFieldTag::PreviousHashLo,
+                            i,
+                            Value::known(
+                                h.to_be_bytes().iter().skip(16).fold(F::ZERO, |acc, byte| {
+                                    acc * F::from(BYTE_POW_BASE) + F::from(*byte as u64)
+                                }),
+                            ),
+                        )
+                    })
+                    .collect_vec()
+                    .as_slice(),
+            );
+            block_data.extend_from_slice(&[(
+                "zero",
+                BlockContextFieldTag::None,
+                0,
+                Value::known(F::ZERO),
+            )]);
+        }
+
+        let mut chain_id_cell = vec![];
+        for (offset, (name, tag, idx, val)) in block_data.into_iter().enumerate() {
+            let absolute_offset = base_offset + offset;
+            self.q_block_table.enable(region, absolute_offset)?;
+            region.assign_advice(
+                || name,
+                self.block_table_blockhash.tag,
+                absolute_offset,
+                || Value::known(F::from(tag as u64)),
+            )?;
+            region.assign_advice(
+                || name,
+                self.block_table_blockhash.index,
+                absolute_offset,
+                || Value::known(F::from(idx as u64)),
+            )?;
+
+            let cell = region.assign_advice(
+                || name,
+                self.block_table_blockhash.value,
+                absolute_offset,
+                || val,
+            )?;
+            if name == "chain_id" {
+                chain_id_cell.push(cell);
+            }
+        }
+
+        Ok(chain_id_cell.pop().unwrap())
+    }
+
     fn assign(
         &self,
         layouter: &mut impl Layouter<F>,
-        public_data: &PublicData,
+        public_data: &PublicData<F>,
+        test_public_data: &Option<PublicData<F>>,
         challenges: &Challenges<Value<F>>,
     ) -> Result<(), Error> {
+        let lt_chip: LtChip<F, 1> = LtChip::construct(self.rlp_is_short);
+        lt_chip.load(layouter)?;
+
         let pi = layouter.assign_region(
             || "region 0",
             |ref mut region| {
+                // avoid the humongous amount of rubbish halo2 errors produced when circuit is
+                // failing
+                for i in 0..BLOCKHASH_TOTAL_ROWS * (PREVIOUS_BLOCKS_NUM + 1) {
```

Comment:
Seems like this may stick around, so would wrap this in a loop iterating over all the columns and set them like `0` like that, because now there's a lot of code here.